### PR TITLE
musl: tighten platforms

### PR DESCRIPTION
**Pull request by alyssais**
###### Description of changes

This will make it possible to check whether we can use pkgsStatic opportunistically, in places like busybox-sandbox-shell, which currently decides not to use pkgsStatic based on a hard-coded set of platforms.

I haven't actually changed busybox-sandbox-shell to use it yet, because that would just cause conflicts with https://github.com/NixOS/nixpkgs/pull/227560, so we can do it later.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/os-specific/linux/musl/default.nix [hunk 1]
```diff
@@ -149,7 +149,13 @@ stdenv.mkDerivation rec {
     homepage    = "https://musl.libc.org/";
     changelog   = "https://git.musl-libc.org/cgit/musl/tree/WHATSNEW?h=v${version}";
     license     = licenses.mit;
-    platforms   = platforms.linux;
+    platforms   = [
+      "aarch64-linux" "armv5tel-linux" "armv6l-linux" "armv7a-linux"
+      "armv7l-linux" "i686-linux" "x86_64-linux" "m68k-linux"
+      "microblaze-linux" "microblazeel-linux" "mipsel-linux"
+      "mips64el-linux" "powerpc64-linux" "powerpc64le-linux"
+      "riscv64-linux" "s390x-linux"
```
Comment (wegank) [inline pkgs/os-specific/linux/musl/default.nix:157]: Not sure about `riscv64-linux`...

Comment (wegank) [reply pkgs/os-specific/linux/musl/default.nix:157]: Ah, nice. Maybe `useMusl = !stdenv.targetPlatform.isRiscV` in `pkgs/stdenv/linux/make-bootstrap-tools.nix` could be updated.

Comment (alyssais) [reply pkgs/os-specific/linux/musl/default.nix:157]: Done.

**Comment by wegank** [comment]
And I think `busybox-sandbox-shell` can be changed now?

**Comment by alyssais** [comment]
> And I think `busybox-sandbox-shell` can be changed now?

In theory, but the check we'd want to use (`lib.meta.availableOn stdenv.hostPlatform pkgsStatic.cc.libc`) currently produces an error because it fails the meta check before getting to our `availableOn` call.  https://github.com/NixOS/nixpkgs/pull/213096 would fix it, so I think it's worth waiting for that.

(We don't want to just check for musl there, because pkgsStatic only means musl on Linux.)

**Comment by alyssais** [comment]
freshBootstrapTools successfully built on riscv64-linux by 0x4a6f (#exotic:nixos.org)

**Comment by NickCao** [comment]
The change looks reasonable to me, but can we somehow subtract the bad platforms instead of listing the good ones?

**Comment by alyssais** [comment]
> The change looks reasonable to me, but can we somehow subtract the bad platforms instead of listing the good ones?

We can, but I don't think we should.  `badPlatforms` is good when you have a package that is generally portable, but for one reason or another is incompatible with a certain platform.  But for stuff like this, where the package needs architecture-specific code for every architecture it supports, it's better to list them explicitly, or we'll end up with false positives any time a new architecture is added to Nixpkgs, and false negatives (when a package is incorrectly marked as being unsupported on a platform) are a lot easier to spot than false positives.

**Pull request merged**